### PR TITLE
[core] Get rid of user-specified refs

### LINKED ISSUE #6900: Get rid of user-specified refs
Parity issue with gl-js:

* https://github.com/mapbox/mapbox-gl-js/issues/1393
* https://github.com/mapbox/mapbox-gl-js/pull/3486

### LINKED ISSUE #7572: Changing ref properties must unlink existing refs
"Ref properties" are 'type', 'source', 'source-layer', 'minzoom', 'maxzoom', 'filter', and 'layout'. Of these, the first three are read-only (except see #7570 for a proposal to make 'source' read-write).

Change any of these properties must at minimum unlink layers that were originally linked via the "ref" mechanism, such that they act independently. Ideally, new possible refs are autodetected -- that's #6900. But short of that, this issue tracks the fact that the existing runtime style API will not behave as expected if you modify, say, filter or layout properties of a layer that was originally a `ref`.

### PR DESCRIPTION
**Pull request by jfirebaugh**
Fixes #6900
Fixes #7572

### CODE REVIEW

File: src/mbgl/style/group_by_layout.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    writer.Uint(static_cast<uint32_t>(layer.baseImpl->visibility));
+    stringify(writer, layer.baseImpl->filter);
+    layer.baseImpl->stringifyLayout(writer);
+    writer.EndArray();
+
+    return s.GetString();
+}
+
+std::vector<std::vector<std::unique_ptr<Layer>>> groupByLayout(std::vector<std::unique_ptr<Layer>> layers) {
+    std::unordered_map<std::string, std::vector<std::unique_ptr<Layer>>> map;
```
Comment (1ec5) [inline src/mbgl/style/group_by_layout.cpp]: `std::vector<std::unique_ptr<Layer>>` is getting repeated quite a bit. Do you think the code would benefit from a `Layout` typedef or somesuch?

File: src/mbgl/util/color.cpp [hunk 2]
```diff
@@ -17,4 +18,12 @@ optional<Color> Color::parse(const std::string& s) {
     }};
 }
 
+std::string Color::stringify() const {
+    return "rgba(" +
```
Comment (1ec5) [inline src/mbgl/util/color.cpp]: Does it matter that a `Color` created via an `hsla()` string would get converted back to JSON as an `rgba()` string?

Comment (jfirebaugh) [inline src/mbgl/util/color.cpp]: Named colors will also come back as `rgba()` strings; there are no guarantees that values are round-tripped exactly. This is consistent with GL JS and the CSSOM.

File: src/mbgl/style/group_by_layout.cpp (hunk 1 above)
Comment (jfirebaugh) [inline src/mbgl/style/group_by_layout.cpp]: This method is short enough that I think it's ok.

**Pull request merged**